lib/api/model/model.dart

```diff
@@ -120,6 +78,41 @@ class User {
   Map<String, dynamic> toJson() => _$UserToJson(this);
 }
 
+/// As in [User.profileData].
+@JsonSerializable(fieldRename: FieldRename.snake)
+class ProfileFieldUserData {
+  final String value;
+  final String? renderedValue;
+
+  ProfileFieldUserData({required this.value, this.renderedValue});
+
+  factory ProfileFieldUserData.fromJson(Map<String, dynamic> json) =>
+      _$ProfileFieldUserDataFromJson(json);
+
+  Map<String, dynamic> toJson() => _$ProfileFieldUserDataToJson(this);
+}
+
+/// As in [User.role].
+@JsonEnum(fieldRename: FieldRename.snake, valueField: "apiValue")
+enum UserRole{
+  owner(apiValue: 100, label: "Owner"),
+  administrator(apiValue: 200, label: "Admin"),
```

Comment:
I've taken your suggestion and moved it into the `widget/profile.dart` UI code for now, and I do agree that there's an argument to keep the `api/model` code related to server interactions, but my initial inclination was to bundle this behavior inside the enum itself (I saw this in https://github.com/zulip/zulip-flutter/blob/main/lib/widgets/compose_box.dart#L92 though to your point that is UI code). It also does appear this is the only place that the role is rendered to a string in the UI, so it's a natural place for someone to come in and look. I guess my inclination comes from assuming if there is some future usage where this enum is rendered to the frontend, we would want to refactor this code out of this specific piece of UI and the logical place that it would end up is in the enum itself.

Reply:
Yeah, it's definitely possible that we'll have some future place we want to show these labels outside the profile screen. I think even then I'll still be inclined to have them live somewhere in UI code — partly for the sake of trying to keep a conceptual separation between `lib/api/` about the server API and `lib/widgets/` about our UI, and partly for the practical reason that I suspect i18n is going to add a bit more complexity to this logic in a way that will feel more natural to handle within UI code.

I think it's fairly likely that the filename `lib/widgets/profile.dart` would still be a good home in that case, in that that other UI would still be some form of presenting details about a user (so arguably a form of their "profile") even though it's not the profile _screen_. But alternatively it might go into some file that's for miscellaneous fragments of UI about a user… perhaps alongside `Avatar` and related widgets.


> to bundle this behavior inside the enum itself (I saw this in https://github.com/zulip/zulip-flutter/blob/main/lib/widgets/compose_box.dart#L92 [[permalink](https://github.com/zulip/zulip-flutter/blob/9e82157747029a55a1012beb69625be36f83ca50/lib/widgets/compose_box.dart#L92)] though to your point that is UI code).

Yeah. That enum ContentValidationError has a bit of a different character in that it itself is basically part of the UI logic, so there's no analogue there of the conceptual separation between server-API code and UI code.